### PR TITLE
pk: implement class ECDHNaive

### PR DESCRIPTION
**Pull request by stepheny**
Add enum _pk.MBEDTLS_ECP_DP_CURVE448 pk.Curve.CURVE448.
Add class pk.ECDHNaive to do key exchange with CURVE25519 or CURVE448.
Exposed mbedtls_ecdh_gen_public, mbedtls_ecdh_compute_shared, as
mbedtls/programs/pkey/ecdh_curve25519.c.

### CODE REVIEW

**Comment by stepheny** [comment]
Still have some doubts on this pr

1. Is the naming ok?
2. curve448 enum
mbedtls had supported curve448 for years. This should works fine but some systems such as debian stretch (debian9, current stable) do ship an ancient version of libmbedtls which do not support curve448. In similar scenario, cython would fail to build.
3. generate, import_peer interface and endianness
Currently a bytes object large enough is used to store public keys with _mpi.mbedtls_mpi_read_binary and _mpi.mbedtls_mpi_write_binary. While mbedtls mpi library "natively" read and write in big endian,
should some glue code be added to support little endian, or use mpi.MPI instead?
4. Should the methods to read private key and public key refactored to ECDHBase?

**Comment by Synss** [comment]
First of all, thank you for your contribution. You are my first contributor for this project!

I will have to review the PR later but for now:

1. ```import_peer()``` could be ```import_peer_public()``` or ```import_peers_public()```? I am not sure either.
2. I have started creating manylinux1 wheels to the project in 0.14.0 to get around the problem of Linux distributions shipping old versions of libmbedtls. My long term goal is to update the wheel to distribute the next long term support version of libmbedtls (2.16.x) and also add ChaCha/Poly. So your contribution with Curve 448 fits very well!
3. Using ```mpi.MPI()``` could be a better solution. I am not sure that ```python-mbedtls``` handles endianness correctly everywhere, however. So you may as well leave it.
4. ```property(_private_key())``` and ```property(_public_key())``` should definitely be in ```ECDHBase```. However, I think that ```_public_key()``` should return an ```ECPoint```.

Generally, could you add tests in ```test_pk.py``` that test every new method at least once? Also, could you add a short sentence in the ChangeLog?

File: src/mbedtls/pk.pyx [hunk 1]
```diff
@@ -81,9 +81,10 @@ class Curve(bytes, enum.Enum):
     BRAINPOOLP384R1 = b'brainpoolP384r1'
     BRAINPOOLP512R1 = b'brainpoolP512r1'
     CURVE25519 = b'curve25519'
-    SECP256K1 = b'secp256k1'
-    SECP224K1 = b'secp224k1'
     SECP192K1 = b'secp192k1'
+    SECP224K1 = b'secp224k1'
+    SECP256K1 = b'secp256k1'
+    CURVE448 = b'curve448'
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: I would rather group the ```CURVE*``` names so that they are easier to find later on.

File: src/mbedtls/pk.pyx [hunk 2]
```diff
@@ -1036,3 +1037,76 @@ cdef class ECDHClient(ECDHBase):
...
+
+    """Naive ECDH key exchange.
+
+    Args:
+        (Curve, optional): b'curve25519' or b'curve448'.
+
+    """
+    def __init__(self, curve=None):
+        super().__init__()
+        self.curve = curve or b'curve25519'
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: I would prefer:
```python
def __init__(self, curve=None):
    if curve is None:
        curve = b"curve25519"
    super().__init__(curve)
    ...
```
as that would be more consistent with the code in ```ECDHBase```.

File: src/mbedtls/pk.pyx [hunk 3]
```diff
@@ -1036,3 +1037,76 @@ cdef class ECDHClient(ECDHBase):
...
+        cdef unsigned char* output = <unsigned char*>malloc(
+            _mpi.MBEDTLS_MPI_MAX_SIZE * sizeof(unsigned char))
+        cdef size_t olen = 0
+        if not output:
+            raise MemoryError()
+        try:
+            check_error(_pk.mbedtls_ecdh_gen_public(
+                &self._ctx.grp, &self._ctx.d, &self._ctx.Q,
+                &_rnd.mbedtls_ctr_drbg_random, &__rng._ctx))
+            n = (_mpi.mbedtls_mpi_bitlen(&self._ctx.Q.X)+7)//8
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: As mentioned in the comments to the PR, using an MPI from the beginning (```_public_key()```) should make the code simpler.

File: src/mbedtls/pk.pyx [hunk 4]
```diff
@@ -1036,3 +1037,76 @@ cdef class ECDHClient(ECDHBase):
...
+
+    @property
+    def _private_key(self):
+        try:
+            return _mpi.from_mpi(&self._ctx.d)
+        except ValueError:
+            return _mpi.MPI()
+
+    @property
+    def _public_key(self):
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: I believe that this should return an ```ECPoint```. In any case, ```_private_key()``` and ```_public_key()``` are actually nice to have. I seem to have overlooked them. Feel free to move them to the base class (as you suggested). Thank you!

File: src/mbedtls/pk.pyx [hunk 5]
```diff
@@ -1036,3 +1037,76 @@ cdef class ECDHClient(ECDHBase):
...
+        try:
+            return _mpi.from_mpi(&self._ctx.Q.X)
+        except ValueError:
+            return _mpi.MPI()
+
+    def generate_secret(self):
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: Should not the call to ```_pk.mbedtls_ecdh_compute_shared()``` be here?

File: Dockerfile [hunk 6]
```diff
@@ -8,7 +8,7 @@ RUN yum -y update \
     && yum -yq install cmake \
     && yum clean all
 COPY ./scripts/ ./scripts/
-RUN ./scripts/install-mbedtls.sh 2.7.9 \
+RUN ./scripts/install-mbedtls.sh 2.14.1 \
```
Comment (Synss) [inline Dockerfile]: You might as well go fo 2.16.0 if you like.

File: src/mbedtls/pk.pyx (hunk 1 above)
Comment (stepheny) [reply src/mbedtls/pk.pyx]: This list is merely `enum mbedtls_ecp_group_id` from `mbedtls/ecp.h`, more or less "sorted by enum value". I think any order should be the same in python enum. You could reorder them anyway convenient.

File: src/mbedtls/pk.pyx (hunk 2 above)
Comment (stepheny) [reply src/mbedtls/pk.pyx]: A little mistake here. It is purposely to override `__init__` from super class, `super().__init__()` should not be called here. I'll remove it later.
`ECDHBase.__init__` invokes `curve_name_to_grp_id` to get the `mbedtls_ecp_group_id` used by `mbedtls_ecp_group_load`. `curve_name_to_grp_id` will search curve name againt `mbedtls_ecp_curve_list`. This is equivalent to do a runtime curve availability check which is very good. But according to ARMmbed/mbedtls#464, CURVE25519 and CURVE448 were still missing and could not be used to do "normal" key exchange.

File: src/mbedtls/pk.pyx (hunk 5 above)
Comment (stepheny) [reply src/mbedtls/pk.pyx]: Updated.

File: src/mbedtls/pk.pyx (hunk 2 above)
Comment (Synss) [reply src/mbedtls/pk.pyx]: Very well.

File: src/mbedtls/pk.pyx [hunk 7]
```diff
@@ -612,7 +612,7 @@ cdef class ECPoint:
             return _pk.mbedtls_ecp_is_zero(&self._ctx) == 1
         elif type(other) is type(self):
             c_other = <ECPoint> other
-            return _pk.mbedtls_ecp_point_cmp(&self._ctx, &c_other._ctx)
+            return _pk.mbedtls_ecp_point_cmp(&self._ctx, &c_other._ctx) == 0
```
Comment (Synss) [inline src/mbedtls/pk.pyx]: 👍

**Comment by Synss** [comment]
Perfect! thank you again.

**Pull request merged**